src/Sarif/EnumeratedArtifact.cs

```diff
@@ -16,44 +16,162 @@ public EnumeratedArtifact(IFileSystem fileSystem)
 
-        private string GetContents()
+        public byte[] Bytes
+        {
+            get => GetArtifactData().bytes;
+            set => this.bytes = value;
+        }
+
+        private (string text, byte[] bytes) GetArtifactData()
         {
-            if (this.contents != null) { return this.contents; }
+            if (this.contents != null)
+            {
+                return (this.contents, bytes: null);
+            }
+
+            if (this.bytes != null)
+            {
+                return (text: null, this.bytes);
+            }
 
-            if (Stream == null && this.contents == null)
+            if (Stream == null && this.contents == null && this.bytes == null)
             {
-                // TBD we actually have no validation URI is non-null yet.
-                contents = Uri!.IsFile
-                    ? FileSystem.FileReadAllText(Uri.LocalPath)
-                    : null;
+                if (Uri == null || (Uri.IsAbsoluteUri && !Uri.IsFile))
+                {
+                    throw new InvalidOperationException();
+                }
 
-                this.sizeInBytes = (long?)this.contents?.Length;
+                // This is our client-side, disk-based file retrieval case.
+                Stream = FileSystem.FileOpenRead(Uri.LocalPath);
+            }
+
+            if (Stream.CanSeek)
+            {
+                RetrieveDataFromSeekableStream();
             }
             else
             {
-                if (Stream.CanSeek) { this.Stream.Seek(0, SeekOrigin.Begin); }
+                RetrieveDataFromNonSeekableStream();
+            }
+
+            Stream = null;
+
+            return (this.contents, this.bytes);
+        }
+
+        private void RetrieveDataFromNonSeekableStream()
+        {
+            bool isText;
+            if (!SupportNonSeekableStreams)
+            {
+                throw new InvalidOperationException("Stream is not seekable. Provide a seekable stream or set the 'SupportNonSeekableStreams' property.");
+            }
+
+            this.bytes = new byte[Stream.Length];
+            int length = this.Stream.Read(this.bytes, 0, this.bytes.Length);
+            isText = FileEncoding.CheckForTextualData(this.bytes, 0, length, out this.encoding);
+
+            if (isText)
+            {
+                // If we have textual data and the encoding was null, we are UTF8
+                // (which will be a perfectly valid encoding for ASCII as well).
+                this.encoding ??= Encoding.UTF8;
+                this.contents = encoding.GetString(this.bytes);
+                this.bytes = null;
+            }
+        }
+
+        private void RetrieveDataFromSeekableStream()
+        {
+            bool isText;
+
+            // Reset to beginning of stream in case caller neglected to do so.
+            this.Stream.Seek(0, SeekOrigin.Begin);
+
+            byte[] header = new byte[4096];
+            int length = this.Stream.Read(header, 0, header.Length);
```

Comment:
I am fully cognizant that this is almost certainly out of scope for what you've set out to accomplish but I'm duty-bound to wonder aloud when I see more synchronous code being crafted that does full file reads: do you see a path to the Task Programming Mode?

Reply:
Sure thing! It is a SMOP. Just a 'small matter of programming'. :)